### PR TITLE
Ensure ByteStrings are compact when calling ToString.

### PR DESCRIPTION
**Pull request by tintoy**
Fixes akkadotnet/akka.net#3147.

### CODE REVIEW

**Comment by Aaronontheweb** [comment]
@tintoy looks like some of the JSON framing specs are failing for Akka.Streams, but the Akka.Persistence issues might be related to this: https://github.com/akkadotnet/akka.net/pull/3149 - should find out once CI has a change to run, but you should take a look at the failing specs for streams.

**Comment by tintoy** [comment]
Yep, if I revert my change, then that test passes (but of course the other tests that I added fail). As far as I can tell, it's the `UnfoldResource` test that seems to be wrong (although I'm not sure why since I find the code a little hard to follow). It specifies a chunk size of `50`, but then expects a string that's only `10` long, which doesn't sound right.

**Comment by tintoy** [comment]
Ok, found the bug - I didn't realise `ByteBuffer` is an alias for `ArraySegment<byte>`, so I was ignoring the other properties of the `ArraySegment` in the case that `IsCompact == true`. I'm just running the remaining streams tests then I'll push a fix.

File: src/core/Akka/Util/ByteString.cs [hunk 1]
```diff
@@ -580,13 +580,12 @@ public IEnumerator<byte> GetEnumerator()
...
-            foreach (var buffer in _buffers)
-            {
-                var part = encoding.GetString(buffer.Array, buffer.Offset, buffer.Count);
-                builder.Append(part);
-            }
-            return builder.ToString();
+            if (IsCompact)
+                return encoding.GetString(_buffers[0].Array, _buffers[0].Offset, _buffers[0].Count);
+
+            byte[] buffer = ToArray();
```
Comment (Horusiath) [inline src/core/Akka/Util/ByteString.cs]: What advantage does that give over using `StringBuilder`? I'm just curious.

Comment (tintoy) [reply src/core/Akka/Util/ByteString.cs]: Hi! The problem is that when you have a `ByteString` composed of several `ByteString`s and some of those `ByteStrings` represent incomplete Unicode sequences (e.g. `"AB"` in Unicode is 4 bytes long, and if you split the `ByteString` after 3 bytes, resulting in a `ByteString` representing 1-and-a-half characters), then the resulting string will be incorrect because some of its components will be incorrect.

Comment (tintoy) [reply src/core/Akka/Util/ByteString.cs]: Most of the existing code and tests used UTF-8, which encodes characters from the ASCII character set using only a single byte (so the issue may not have come up before). As soon as I switched to Unicode it gave wrong results unless I called `.Compact()` before calling `.ToString()`.

Comment (tintoy) [reply src/core/Akka/Util/ByteString.cs]: [Example](https://github.com/akkadotnet/akka.net/pull/3148/files#diff-07edeee541b3c2f1fbe2a9aab5458214R135)

**Comment by Aaronontheweb** [comment]
I like that there are multiple test cases for UTF8, Unicode, et al here. Looks like a good change to me. We'll get this into 1.3.2
